src/Http/Environment.php

```diff
@@ -0,0 +1,989 @@
+    protected $host;
+
+    /**
+     * Whether the HTTPS protocol is used
+     *
+     * @var bool
+     */
+    protected $https;
+
+    /**
+     * Sanitized `$_SERVER` data
+     *
+     * @var array
+     */
+    protected $info;
+
+    /**
+     * Current server's IP address
+     *
+     * @var string
+     */
+    protected $ip;
+
+    /**
+     * Whether the site is behind a reverse proxy
+     *
+     * @var bool
+     */
+    protected $isBehindProxy;
+
+    /**
+     * URI path to the base
+     *
+     * @var string
+     */
+    protected $path;
+
+    /**
+     * Port number in the site URL
+     *
+     * @var int|null
+     */
+    protected $port;
+
+    /**
+     * Intermediary value of the port
+     * extracted from the host name
+     *
+     * @var int|null
+     */
+    protected $portInHost;
+
+    /**
+     * Array with path and query
+     * from the `REQUEST_URI`
+     *
+     * @var array
+     */
+    protected $requestUri;
+
+    /**
+     * Root directory for environment configs
+     *
+     * @var string
+     */
+    protected $root;
+
+    /**
+     * Path to the php script within the
+     * document root without the
+     * filename of the script
+     *
+     * @var string
+     */
+    protected $scriptPath;
+
+    /**
+     * Full base URL
+     *
+     * @var string
+     */
+    protected $url;
+
+    /**
+     * Class constructor
+     *
+     * @param array|null $options
+     * @param array|null $info Optional override for `$_SERVER`
+     */
+    public function __construct(?array $options = null, ?array $info = null)
+    {
+        $this->detect($options, $info);
+    }
+
+    /**
+     * Returns the server's IP address
+     *
+     * @see static::ip
+     * @return string|null
+     */
+    public function address(): ?string
+    {
+        return $this->ip();
+    }
+
+    /**
+     * Checks if the request is being served by the CLI
+     *
+     * @return bool
+     */
+    public function cli(): bool
+    {
+        return $this->cli;
+    }
+
+    /**
+     * Sanitizes the server info and detects
+     * all relevant parts. This can be called
+     * again at a later point to overwrite all
+     * the stored information and re-detect the
+     * environment if necessary.
+     *
+     * @param array|null $options
+     * @param array|null $info Optional override for `$_SERVER`
+     * @return array
+     */
+    public function detect(array $options = null, array $info = null): array
+    {
+        $info ??= $_SERVER;
+        $options = array_merge([
+            'cli'     => null,
+            'root'    => null,
+            'allowed' => null
+        ], $options ?? []);
+
+        $this->info          = $this->sanitize($info);
+        $this->cli           = $this->detectCli($options['cli']);
+        $this->ip            = $this->detectIp();
+        $this->host          = null;
+        $this->https         = false;
+        $this->isBehindProxy = $this->detectIfIsBehindProxy();
```

Comment:
This line allows visitors/attackers to completely control the `isBehindProxy` flag by setting the forwarded headers, even on setups where the forwarded headers are not used in the end (disabled by the `allowed` config).

I think we should base this value on the fact whether the forwarded host is actually used for the environment detection. For example we could set it in `detectAuto()`.

Reply:
I've committed a possible fix, please check if you agree with it.